diff --git a/schemas/ISO-IEC29500-4_2016/wml.xsd b/schemas/ISO-IEC29500-4_2016/wml.xsd
index 6fa33d9..e243504 100644
--- a/schemas/ISO-IEC29500-4_2016/wml.xsd
+++ b/schemas/ISO-IEC29500-4_2016/wml.xsd
@@ -3044,6 +3044,7 @@
       <xsd:element name="targetScreenSz" type="CT_TargetScreenSz" minOccurs="0"/>
       <xsd:element name="saveSmartTagsAsXml" type="CT_OnOff" minOccurs="0"/>
     </xsd:sequence>
+        <xsd:attribute ref="mc:Ignorable" use="optional" />
   </xsd:complexType>
   <xsd:simpleType name="ST_FrameScrollbar">
     <xsd:restriction base="xsd:string">
@@ -3194,6 +3195,7 @@
       <xsd:element name="num" type="CT_Num" minOccurs="0" maxOccurs="unbounded"/>
       <xsd:element name="numIdMacAtCleanup" type="CT_DecimalNumber" minOccurs="0"/>
     </xsd:sequence>
+        <xsd:attribute ref="mc:Ignorable" use="optional" />
   </xsd:complexType>
   <xsd:simpleType name="ST_TblStyleOverrideType">
     <xsd:restriction base="xsd:string">
@@ -3285,6 +3287,7 @@
       <xsd:element name="latentStyles" type="CT_LatentStyles" minOccurs="0" maxOccurs="1"/>
       <xsd:element name="style" type="CT_Style" minOccurs="0" maxOccurs="unbounded"/>
     </xsd:sequence>
+        <xsd:attribute ref="mc:Ignorable" use="optional" />
   </xsd:complexType>
   <xsd:complexType name="CT_Panose">
     <xsd:attribute name="val" type="s:ST_Panose" use="required"/>
@@ -3348,6 +3351,7 @@
     <xsd:sequence>
       <xsd:element name="font" type="CT_Font" minOccurs="0" maxOccurs="unbounded"/>
     </xsd:sequence>
+        <xsd:attribute ref="mc:Ignorable" use="optional" />
   </xsd:complexType>
   <xsd:complexType name="CT_DivBdr">
     <xsd:sequence>
